### PR TITLE
New module for managing volumes in Vexata storage arrays.

### PR DESCRIPTION
**Pull request by vexata**
##### SUMMARY
Adding new module for managing volumes in Vexata storage arrays.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
vexata_volume

##### ANSIBLE VERSION

```ansible 2.8.0.dev0 (devel 871dd77f2b) last updated 2018/10/15 11:57:16 (GMT -700)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/test/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/test/dev/ansible/lib/ansible
  executable location = /home/test/dev/ansible/bin/ansible
  python version = 2.7.6 (default, Nov 23 2017, 15:49:48) [GCC 4.8.4]
```

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/module_utils/vexata.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Ansible, but is an independent component.
```
Comment (Akasurde) [inline lib/ansible/module_utils/vexata.py]: Replace line 2 - 29 with 

```
# Copyright: (c) 2018, Sandeep Kasargod <sandeep@vexata.com>
# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
```

File: lib/ansible/module_utils/vexata.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+def get_array(module):
+    """Return storage array object or fail"""
+    global VXOS_VERSION
+    array = module.params['array']
+    user = module.params.get('user', None)
+    password = module.params.get('password', None)
+
+    if user and password:
+        system = VexataAPIProxy(array, user, password, verify_cert=False)
```
Comment (Akasurde) [inline lib/ansible/module_utils/vexata.py]: Make `verify_cert` configurable. You can take a look at [this](https://github.com/ansible/ansible/blob/959395f4b40a4f9e44a4bce890f633f8364c43a6/lib/ansible/module_utils/vmware.py#L466)

File: lib/ansible/module_utils/vexata.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    elif environ.get('VEXATA_USER') and environ.get('VEXATA_PASSWORD'):
+        user = environ.get('VEXATA_USER')
```
Comment (Akasurde) [inline lib/ansible/module_utils/vexata.py]: You can use [`env_fallback`](https://github.com/ansible/ansible/blob/8f41270a010c00d058c70bdccdc611df8b454139/lib/ansible/module_utils/basic.py#L726)

File: lib/ansible/module_utils/vexata.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                             'the module as arguments or by setting the '
+                             'VEXATA_USER and VEXATA_PASSWORD environment variables.')
+    try:
+        if system.test_connection():
+            VXOS_VERSION = get_version(system.iocs())
+            return system
+        else:
+            module.fail_json(msg='Test connection to array failed.')
+    except Exception as e:
+        module.fail_json(msg='Vexata API access failed: {0}'.format(str(e)))
```
Comment (Akasurde) [inline lib/ansible/module_utils/vexata.py]:
```suggestion
        from ansible.module_utils._text import to_native
        ...
        module.fail_json(msg='Vexata API access failed: {0}'.format(to_native(e)))
```

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# (c) 2018, Sandeep Kasargod (sandeep@vexata.com)
```
Comment (Akasurde) [inline lib/ansible/modules/storage/vexata/vexata_volume.py]:
```suggestion
# Copyright: (c) 2018, Sandeep Kasargod (sandeep@vexata.com)
```

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: vexata_volume
+version_added: 2.8
+short_description: Manage volumes on Vexata VX100 storage arrays.
```
Comment (Akasurde) [inline lib/ansible/modules/storage/vexata/vexata_volume.py]:
```suggestion
short_description: Manage volumes on Vexata VX100 storage arrays
```

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+description:
+    - Create, deletes or extend volumes on a Vexata VX100 array.
+author: Sandeep Kasargod
```
Comment (Akasurde) [inline lib/ansible/modules/storage/vexata/vexata_volume.py]:
```suggestion
author:
- Sandeep Kasargod (@vexata)
```

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+            module.params['name'],
+            'Ansible volume',
+            size)
+        if vol:
+            module.log(msg='Created volume {0}'.format(vol['id']))
+            changed = True
+        else:
+            module.fail_json(msg='Volume create failed.')
+    except Exception:
+        pass
```
Comment (Akasurde) [inline lib/ansible/modules/storage/vexata/vexata_volume.py:112]: Passing here for exception will be result in unexpected results.

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    module.exit_json(changed=changed)
+
+
+def update_volume(module, array, volume):
+    """Expand the volume size."""
+    changed = False
+    size = size_to_MiB(module.params['size'])
```
Comment (Akasurde) [inline lib/ansible/modules/storage/vexata/vexata_volume.py:121]: `size` is optional parameter so adding check against `None` would help here.

```python
if not module.params.get('size', False):
    module.fail_json(msg="Size is required to update volume")

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+            state=dict(default='present', choices=['present', 'absent']),
+            size=dict()
+        )
+    )
+
+    module = AnsibleModule(arg_spec,
+                           supports_check_mode=True,
+                           required_together=required_together())
+
+    if not HAS_VEXATAPI:
```
Comment (Akasurde) [inline lib/ansible/modules/storage/vexata/vexata_volume.py]: This check can be also moved to `module_utils`.

File: lib/ansible/modules/storage/vexata/vexata_volume.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+        module.fail_json(msg='vexatapi library is required for this module. '
+                             'To install, use `pip install vexatapi`')
+
+    state = module.params['state']
+    size = module.params['size']
+    if size and size_to_MiB(size) < 0:
```
Comment (Akasurde) [inline lib/ansible/modules/storage/vexata/vexata_volume.py:178]: `Size` is not required for `delete_volume` so would force user to specify while deleting volume even if it is not required.

File: lib/ansible/utils/module_docs_fragments/vexata.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
+#
+# (c) 2018, Sandeep Kasargod <sandeep@vexata.com>
```
Comment (Akasurde) [inline lib/ansible/utils/module_docs_fragments/vexata.py]: Line 2-17 

```
# Copyright: (c) 2018, Sandeep Kasargod <sandeep@vexata.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

**Review by Akasurde — CHANGES_REQUESTED** [review]
Could you please add test case for this change ? Thanks.

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 8 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py:112]: If there is an exception other than volume create failing, changed remains False so there is no unexpected result.

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 11 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py:178]: - Yes, it is not required for deleting the volume.
- But size is not required in the argument spec and and this code does not force the user to specify it for deleting the volume

File: lib/ansible/module_utils/vexata.py (hunk 1 above)
Comment (vexata) [reply lib/ansible/module_utils/vexata.py]: Fixed copyright. Will submit tests in separate commit, have been waiting for review as there are several other modules in the pipeline

File: lib/ansible/module_utils/vexata.py (hunk 2 above)
Comment (vexata) [reply lib/ansible/module_utils/vexata.py]: Done. For now the storage array default remains False.

File: lib/ansible/module_utils/vexata.py (hunk 3 above)
Comment (vexata) [reply lib/ansible/module_utils/vexata.py]: Done

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 7 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py]: Done

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 9 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py:121]: Done

File: lib/ansible/utils/module_docs_fragments/vexata.py (hunk 12 above)
Comment (vexata) [reply lib/ansible/utils/module_docs_fragments/vexata.py]: Done

File: lib/ansible/modules/storage/vexata/vexata_volume.py (hunk 10 above)
Comment (vexata) [reply lib/ansible/modules/storage/vexata/vexata_volume.py]: Done

**Comment by vexata** [comment]
Messed up this PR, closing it and opened [PR 49006](https://github.com/ansible/ansible/pull/49006) with the changes

**Pull request closed without merging**